### ISSUE TITLE
Cannot find NORMALIZATION OPERATION in REINFORCE++ implementation

### ISSUE THREAD

**Issue by shayuewt** [OpenRLHF/OpenRLHF#924]
Hey, `REINFORCE++` is such a great job and I try to implement it in our RL Framework.

When I want to take the `OpenRLHF` as reference and read the source code, I'm sorry that I couldn't find the **normalization operation for rewards and advantages** as described in paper.

If I am not mistaken, `advantage_estimator ==reinforce` stands for choosing algorithm `reinforce++` in the implemenatition. When I check the code at `openrlhf/trainer/ppo_utils/experience_maker.py`, I couldn't find the normalization for both rewards and advantages.

```python
def compute_advantages_and_returns(
        self, experiences: List[Experience], **kwargs
    ) -> Tuple[List[Experience], List[torch.Tensor]]:
    ...
    # reward shaping, HERE NOT DO NORMALIZATION to rewards FOR "reinforce"
      if args.advantage_estimator == "rloo":
          rewards = torch.cat(rewards).reshape(-1, args.n_samples_per_prompt).to(device="cuda")
          baseline = (rewards.sum(-1, keepdim=True) - rewards) / (args.n_samples_per_prompt - 1)
          rewards = rewards - baseline
          rewards = rewards.flatten().to(device="cpu").chunk(len(experiences))
      elif args.advantage_estimator in ["reinforce_baseline", "dr_grpo"]:
          # REINFORCE++-baseline and Dr. GRPO removed the `/std` in GRPO as `/ std` is not needed in RL variance reduction theory.
          # And `k3 KL` has a larger variance than `k1 KL` under a categorical distribution.
          rewards = torch.cat(rewards).reshape(-1, args.n_samples_per_prompt).to(device="cuda")
          rewards = rewards - rewards.mean(-1, keepdim=True)
          rewards = rewards.reshape(-1).to(device="cpu").chunk(len(experiences))
      elif args.advantage_estimator == "group_norm":
          rewards = torch.cat(rewards).reshape(-1, args.n_samples_per_prompt).to(device="cuda")
          rewards = (rewards - rewards.mean(-1, keepdim=True)) / (rewards.std(-1, keepdim=True) + 1e-9)
          rewards = rewards.reshape(-1).to(device="cpu").chunk(len(experiences))
   ...
```

And same as in function `get_cumulative_returns`

```python
@torch.no_grad()
def get_cumulative_returns(
    self,
    rewards: torch.Tensor,
    action_mask: torch.Tensor,
    gamma: float,
) -> Tuple[torch.Tensor, torch.Tensor]:

    if isinstance(rewards, list):
        # packing samples
        # TODO: this is slow...
        returns = []
        for r in rewards:
            ret = self.get_cumulative_returns(r.unsqueeze(0), action_mask, gamma)
            returns.append(ret.squeeze(0))
        return returns

    response_length = rewards.size(1)
    returns = torch.zeros_like(rewards)
    cumulative_return = torch.zeros(rewards.size(0), device=rewards.device)

    # Mask invalid responses if action_mask is provided
    if action_mask is not None:
        rewards = action_mask * rewards

    # Calculate returns by accumulating discounted rewards
    for t in reversed(range(response_length)):
        cumulative_return = rewards[:, t] + gamma * cumulative_return
        returns[:, t] = cumulative_return

    return returns
```

If I have misunderstood or omitted something, I would really appreciate it if you could tell me. Thank you very much!

**Comment by millioniron** [comment]
replay_buffer

**Comment by shayuewt** [comment]
I find it in `openrlhf/trainer/ray/ppo_actor.py`

```python
  if self.args.advantage_estimator not in ["group_norm", "dr_grpo"]:
      self.replay_buffer.normalize(
          self.strategy, "advantages", divide_by_std=not self.args.no_advantage_std_norm
      )
```

Thanks for your kindness.
